### PR TITLE
Added API support for Import ASG to ocean cluster and VNG, Import EKS…

### PR DESCRIPTION
**Pull request by janetlinkiruba**
… Vng to ocean Vng, Get Elastilog

_This is the default pull request template. You can customize it by adding a `pull_request_template.md` at the root of your repo or inside the `.github` folder._

# Jira Ticket

_Include a link to your Jira Ticket_
_Example: [JIRAISS-1234](https://spotinst.atlassian.net/browse/JIRAISS-1234)_

# Demo

_Please add a recording of the feature/bug fix in work. if you added new routes, the recording should show the request and response for each new/changed route_

# Checklist:
- [ ] I have filled relevant self assessment ([NodeJS](https://docs.google.com/forms/d/e/1FAIpQLSfl14u9AOBAmxVJ272tvO7XNuXE-EMvWaGcaRZalu1UAKB7RA/viewform), [Frontend](https://docs.google.com/forms/d/e/1FAIpQLSdiBPNKH81w_EkavihVL8Uwb0j7tP8PwJmLFYm2nCOQxz-1qw/viewform), [Backend](https://docs.google.com/forms/d/e/1FAIpQLSed_PsTJ5-XIWkFL6BSDE2AQRBVPmwc3PAmHZkUn-erzVI37Q/viewform?usp=sf_link))
- [ ] I have run ESlint on my changes and fixed all warnings and errors (**NodeJS & Frontend Services**)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have validated all the requirements in the Jira task were answered
- [ ] I have all neccessary approvals for the design/mini design of this task
- [ ] I have approved the API changes and granular permission patterns (documentation subtask) (**For public services only**)

### CODE REVIEW

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesClusterUsageExample.java [hunk 1]
```diff
@@ -212,6 +221,40 @@ private static GetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStatus
...
     }
 
+    private static OceanK8sCluster importASGToOceanCluster(SpotOceanK8sClusterClient client, String autoscalingGroupName, String region) {
+
+        System.out.println(String.format("Import ASG to ocean cluster. ASG name: %s", autoscalingGroupName));
+        ImportAsgToClusterInstanceTypes.Builder instanceTypesBuilder = ImportAsgToClusterInstanceTypes.Builder.get();
+        List<String> instancetypes = Arrays.asList("c4.xlarge");
+        ImportAsgToClusterInstanceTypes instanceTypes = instanceTypesBuilder.setInstanceTypes(instancetypes).build();
+
+        System.out.println(instanceTypes.toString());
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesClusterUsageExample.java:231]: This statement can be removed.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesClusterUsageExample.java [hunk 2]
```diff
@@ -212,6 +221,40 @@ private static GetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStatus
         return clusterHeartBeatStatus;
     }
 
+    private static OceanK8sCluster importASGToOceanCluster(SpotOceanK8sClusterClient client, String autoscalingGroupName, String region) {
+
+        System.out.println(String.format("Import ASG to ocean cluster. ASG name: %s", autoscalingGroupName));
+        ImportAsgToClusterInstanceTypes.Builder instanceTypesBuilder = ImportAsgToClusterInstanceTypes.Builder.get();
+        List<String> instancetypes = Arrays.asList("c4.xlarge");
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesClusterUsageExample.java:228]: Please add multiple instance types.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesClusterUsageExample.java [hunk 3]
```diff
@@ -212,6 +221,40 @@ private static GetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStatus
...
+        ImportAsgToClusterInstanceTypes instanceTypes = instanceTypesBuilder.setInstanceTypes(instancetypes).build();
+
+        System.out.println(instanceTypes.toString());
+
+        OceanK8sCluster importASGToOceanCluster = client.importASGToOceanCluster(instanceTypes, autoscalingGroupName, region);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesClusterUsageExample.java:233]: Rename variable importASGToOceanCluster to asgConfigToImport

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 4]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
     }
+
+    private static K8sVngSpec importASGToVng(K8sVngClient client, String autoScalingGroupName, String oceanId) {
+        System.out.println("-------------------------Import ASG to virtual node group------------------------");
+
+        //Build labels
+        K8sVngLabels.Builder labelsBuilder = K8sVngLabels.Builder.get();
+
+        K8sVngLabels       label     = labelsBuilder.setKey("Creator").setValue("testingSdkOcean").build();
+        List<K8sVngLabels> labelsList = Collections.singletonList(label);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:306]: Shouldn't it be just list.add operation.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 5]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+
+        //Build taints
+        K8sVngTaints.Builder taintsBuilder = K8sVngTaints.Builder.get();
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:309]: Shouldn't it be just list.add operation.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 6]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+
+        K8sVngTaints       taints     = taintsBuilder.setKey("Creator").setValue("testingSdkOcean").setEffect("NoSchedule").build();
+        List<K8sVngTaints> taintsList = Collections.singletonList(taints);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:312]: Shouldn't it be just list.add operation.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 7]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+        //Build labels
+        K8sVngLabels.Builder labelsBuilder = K8sVngLabels.Builder.get();
+
+        K8sVngLabels       label     = labelsBuilder.setKey("Creator").setValue("testingSdkOcean").build();
+        List<K8sVngLabels> labelsList = Collections.singletonList(label);
+
+        //Build taints
+        K8sVngTaints.Builder taintsBuilder = K8sVngTaints.Builder.get();
+
+        K8sVngTaints       taints     = taintsBuilder.setKey("Creator").setValue("testingSdkOcean").setEffect("NoSchedule").build();
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:311]: in example, set key as "app" value as "frontend"

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 8]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+        List<K8sVngTaints> taintsList = Collections.singletonList(taints);
+
+        // Build launch spec
+        K8sVngSpec.Builder asgImportRequestBuilder = K8sVngSpec.Builder.get();
+        K8sVngSpec asgImportRequest               =
+                asgImportRequestBuilder.setName("specialty.nodes.spotk8s.com").setLabels(labelsList).setTaints(taintsList).build();
+
+        //Convert launch spec to API object json
+        System.out.println(asgImportRequest.toString());
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:320]: Can be removed.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 9]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+        // Import ASG to VNG
+        K8sVngSpec      importASGResponse = client.importASGToVng(asgImportRequest, autoScalingGroupName, oceanId);
+
+        System.out.println(String.format("Response: %s", importASGResponse.toString()));
+
+        return importASGResponse;
+    }
+
+    private static K8sVngSpec importClusterVngToOceanVng(K8sVngClient client) {
+        System.out.println("-------------------------Import ASG to virtual node group------------------------");
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:331]: Incorrect message.

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 10]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+
+        //Build labels
+        K8sVngLabels.Builder labelsBuilder = K8sVngLabels.Builder.get();
+
+        K8sVngLabels       label     = labelsBuilder.setKey("Creator").setValue("testingSdkOcean").build();
+        List<K8sVngLabels> labelsList = Collections.singletonList(label);
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:337]: Create multiple labels and add to list using list.add

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 11]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+                asgImportRequestBuilder.setName("specialty.nodes.spotk8s.com").setLabels(labelsList).build();
+
+        //Build launch spec creation request
+        K8sImportClusterVngToOceanVngRequest.Builder vngImportRequestBuilder = K8sImportClusterVngToOceanVngRequest.Builder.get();
+        K8sImportClusterVngToOceanVngRequest        importRequest           = vngImportRequestBuilder.setVngLaunchSpec(asgImportRequest).setAccountId(accountId).setEksClusterName("test").setEksNodeGroupName("test_vng").setOceanId(oceanId).build();
+
+        //Convert launch spec to API object json
+        System.out.println(importRequest.toJson());
+
+        // Import ASG to VNG
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:352]: Incorrect message

File: src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java [hunk 12]
```diff
@@ -287,4 +295,65 @@ private static void deleteK8sVng(K8sVngClient client, String launchSpecId) {
...
+
+        //Build labels
+        K8sVngLabels.Builder labelsBuilder = K8sVngLabels.Builder.get();
+
+        K8sVngLabels       label     = labelsBuilder.setKey("Creator").setValue("testingSdkOcean").build();
+        List<K8sVngLabels> labelsList = Collections.singletonList(label);
+
+
+        // Build launch spec
+        ImportEKSK8sVngSpec.Builder asgImportRequestBuilder = ImportEKSK8sVngSpec.Builder.get();
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/example/ocean/kubernetes/OceanKubernetesVngSpecUsageExample.java:341]: variable names are referring asg throughout this function. It should refer Vng.

File: src/main/java/com/spotinst/sdkjava/model/OceanK8sConverter.java [hunk 13]
```diff
@@ -698,6 +712,10 @@ private static ClusterInstanceTypes toBl(ApiClusterInstanceTypes apiInstanceType
             }
             if (apiInstanceTypes.isWhitelistSet()) {
                 apiInstanceTypes.setWhitelist(apiInstanceTypes.getWhitelist());
+                instanceTypesBuilder.setBlacklist(apiInstanceTypes.getBlacklist());
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/OceanK8sConverter.java:715]: Please recheck at this place.

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterClient.java [hunk 14]
```diff
@@ -159,4 +160,50 @@ public GetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStatus(String
         return getK8sClusterHeartBeatStatus;
 
     }
+
+    public OceanK8sCluster importASGToOceanCluster(ImportAsgToClusterInstanceTypes importAsg, String autoScalingGroupName, String region) {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterClient.java:164]: please change importAsg to instanceTypes

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterClient.java [hunk 15]
```diff
@@ -159,4 +160,50 @@ public GetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStatus(String
...
+
+        OceanK8sCluster importASGToCluster;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterClient.java:166]: Please change importASGToCluster to importedClusterConfiguration

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterClient.java [hunk 16]
```diff
@@ -159,4 +160,50 @@ public GetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStatus(String
...
+                    "Error encountered while attempting to import ASG to ocean cluster. Code: %s. Message: %s.",
+                    httpException.getCode(), httpException.getMessage()));
+            throw new SpotinstHttpException(httpException.getMessage());
+        }
+
+        return importASGToCluster;
+
+    }
+
+    public List<K8sClusterFetchElastilogResponse> fetchElastilog(K8sClusterFetchElastilogRequest elastigroupGetElastilogRequest, String clusterId) {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterClient.java:187]: Variable name "elastigroupGetElastilogRequest" should be changed.

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterService.java [hunk 17]
```diff
@@ -243,4 +248,120 @@ public static ApiGetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStat
         return getClusterHeartBeatResponse;
 
     }
+
+    public static ApiOceanK8sCluster importASGToOceanCluster(ImportAsgToClusterInstanceTypes importRequest, String autoScalingGroupName, String region, String authToken, String account) {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterService.java:252]: Change variable name to "importRequest"

File: src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterService.java [hunk 18]
```diff
@@ -243,4 +248,120 @@ public static ApiGetK8sClusterHeartBeatStatusResponse getK8sClusterHeartBeatStat
...
+        // Add region Query param
+        if (region != null) {
+            queryParams.put("region", region);
+        }
+
+        // Get the headers
+        Map<String, String> headers = buildHeaders(authToken);
+
+        // Write to json
+        Map<String, ImportAsgToClusterInstanceTypes> importClusterRequest = new HashMap<>();
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/SpotOceanK8sClusterService.java:282]: This should be a POST request with launchSpec

File: src/main/java/com/spotinst/sdkjava/model/api/ocean/kubernetes/ApiImportEKSK8sVngSpec.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonFilter("PartialUpdateEntityFilter")
+public class ApiImportEKSK8sVngSpec implements IPartialUpdateEntity {
+    @JsonIgnore
+    private Set<String>                      isSet;
+    private String                           name;
+    private List<ApiK8sVngLabels>			 labels;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/api/ocean/kubernetes/ApiImportEKSK8sVngSpec.java:20]: Should be changed to ApiK8sVngLabel

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/kubernetes/ImportEKSK8sVngSpec.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ImportEKSK8sVngSpec {
+    @JsonIgnore
+    private Set<String>                     isSet;
+    private String                          name;
+	private List<K8sVngLabels>			 	labels;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/kubernetes/ImportEKSK8sVngSpec.java:13]: Align.
Should be changed to K8sVngLabel

File: src/main/java/com/spotinst/sdkjava/model/requests/ocean/kubernetes/ImportASGToK8sClusterRequest.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+import com.spotinst.sdkjava.model.OceanK8sConverter;
+import com.spotinst.sdkjava.model.api.ocean.kubernetes.ApiImportAsgToClusterInstanceTypes;
+import com.spotinst.sdkjava.model.bl.ocean.kubernetes.ImportAsgToClusterInstanceTypes;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.*;
+
+public class ImportASGToK8sClusterRequest {
+    //region Members
+    @JsonProperty("cluster")
+    private ImportAsgToClusterInstanceTypes importAsgToClusterInstanceTypes;
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/requests/ocean/kubernetes/ImportASGToK8sClusterRequest.java:13]: variable name can be just instanceTypes

File: src/main/java/com/spotinst/sdkjava/model/requests/ocean/kubernetes/ImportASGToK8sClusterRequest.java [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+import com.spotinst.sdkjava.client.rest.JsonMapper;
+import com.spotinst.sdkjava.model.OceanK8sConverter;
+import com.spotinst.sdkjava.model.api.ocean.kubernetes.ApiImportAsgToClusterInstanceTypes;
+import com.spotinst.sdkjava.model.bl.ocean.kubernetes.ImportAsgToClusterInstanceTypes;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.*;
+
+public class ImportASGToK8sClusterRequest {
+    //region Members
+    @JsonProperty("cluster")
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/requests/ocean/kubernetes/ImportASGToK8sClusterRequest.java:12]: shouldn't this be launchSpec?

File: src/main/java/com/spotinst/sdkjava/model/service/ocean/kubernetes/K8sVngSpecService.java [hunk 23]
```diff
@@ -203,4 +206,118 @@ public static ApiK8sVngSpec getK8sVng(String oceanLaunchSpecId, String authToken
...
+                getCastedResponse(response, K8sVngApiResponse.class);
+
+        if (importASGToVngResponse.getResponse().getCount() > 0) {
+            retVal = importASGToVngResponse.getResponse().getItems().get(0);
+        }
+
+        return retVal;
+    }
+
+    public static ApiK8sVngSpec importCluterVngToOceanVng(K8sImportClusterVngToOceanVngRequest importClusterVngRequest, String authToken) throws SpotinstHttpException {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/service/ocean/kubernetes/K8sVngSpecService.java:261]: change importCluterVngToOceanVng  to importClusterVngToOceanVng

**Review by anuragsharma-123 — CHANGES_REQUESTED** [review]
Please rework on the review comments

File: src/main/java/com/spotinst/sdkjava/model/OceanK8sConverter.java (hunk 13 above)
Comment (janetlinkiruba) [reply src/main/java/com/spotinst/sdkjava/model/OceanK8sConverter.java:715]: Fixed the issue

File: src/main/java/com/spotinst/sdkjava/model/bl/ocean/kubernetes/ImportAsgToClusterInstanceTypes.java [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class ImportAsgToClusterInstanceTypes {
```
Comment (anuragsharma-123) [inline src/main/java/com/spotinst/sdkjava/model/bl/ocean/kubernetes/ImportAsgToClusterInstanceTypes.java:13]: ImportAsgToClusterConfiguration

**Review by anuragsharma-123 — APPROVED** [review]

**Pull request merged**